x-pack/plugins/cloud_security_posture/server/routes/graph/v1.ts

```diff
@@ -0,0 +1,347 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { castArray } from 'lodash';
+import type { Logger, IScopedClusterClient } from '@kbn/core/server';
+import type {
+  EdgeDataModel,
+  NodeDataModel,
+  EntityNodeDataModel,
+  LabelNodeDataModel,
+  GroupNodeDataModel,
+  NodeShape,
+} from '@kbn/cloud-security-posture-common/types/graph/latest';
+import type { EsqlToRecords } from '@elastic/elasticsearch/lib/helpers';
+import type { Writeable } from '@kbn/zod';
+import type { GraphContextServices, GraphContext } from './types';
+
+interface GraphEdge {
+  badge: number;
+  ips: string[];
+  hosts: string[];
+  users: string[];
+  actorIds: string[] | string;
+  action: string;
+  targetIds: string[] | string;
+  eventOutcome: string;
+  isAlert: boolean;
+}
+
+export const getGraph = async (
+  services: GraphContextServices,
+  query: {
+    actorIds: string[];
+    eventIds: string[];
+    spaceId?: string;
+    start: string | number;
+    end: string | number;
+  }
+): Promise<{
+  nodes: NodeDataModel[];
+  edges: EdgeDataModel[];
+}> => {
+  const { esClient, logger } = services;
+  const { actorIds, eventIds, spaceId = 'default', start, end } = query;
+
+  logger.debug(
+    `Fetching graph for [eventIds: ${eventIds.join(', ')}] [actorIds: ${actorIds.join(
+      ', '
+    )}] in [spaceId: ${spaceId}]`
+  );
+
+  const results = await fetchGraph({ esClient, logger, start, end, eventIds, actorIds });
+
+  // Convert results into set of nodes and edges
+  const graphContext = parseRecords(logger, results.records);
+
+  return { nodes: graphContext.nodes, edges: graphContext.edges };
+};
+
+interface ParseContext {
+  nodesMap: Record<string, NodeDataModel>;
+  edgeLabelsNodes: Record<string, string[]>;
+  edgesMap: Record<string, EdgeDataModel>;
+}
+
+const parseRecords = (logger: Logger, records: GraphEdge[]): GraphContext => {
+  const nodesMap: Record<string, NodeDataModel> = {};
+  const edgeLabelsNodes: Record<string, string[]> = {};
+  const edgesMap: Record<string, EdgeDataModel> = {};
+
+  logger.trace(`Parsing records [length: ${records.length}]`);
+
+  createNodes(logger, records, { nodesMap, edgeLabelsNodes });
+  createEdgesAndGroups(logger, { edgeLabelsNodes, edgesMap, nodesMap });
+
+  logger.trace(
+    `Parsed [nodes: ${Object.keys(nodesMap).length}, edges: ${Object.keys(edgesMap).length}]`
+  );
+
+  // Sort groups to be first (fixes minor layout issue)
+  const nodes = Object.values(nodesMap).sort((a, b) => {
+    if (a.shape === b.shape) {
+      return 0;
+    } else if (a.shape === 'group') {
+      return -1;
+    } else if (b.shape === 'group') {
+      return 1;
+    }
+
+    return 0;
+  });
```

Comment:
perf nit:

If that's going to parse hundreds or even thousands of records I would suggest using filtering and concatenating, instead of sorting, as filter and concatenating has a O(N) time complexity, while sorting has  O(N log N)

```suggestion
  // Move 'group' nodes to the beginning (fixes minor layout issue)
  const nodesArray = Object.values(nodesMap);
  const groupNodes = nodesArray.filter((node) => node.shape === 'group');
  const otherNodes = nodesArray.filter((node) => node.shape !== 'group');
  const nodes = [...groupNodes, ...otherNodes];
```